File: src/main/kotlin/java/_03_staying_immutable/generic/Grid.java
package java._03_staying_immutable.generic;

import java.util.ArrayList;

class Grid<T> {

    private int rows;
    private int cols;

    private ArrayList<T> items;

    public Grid(int rows, int cols, T initialValue) {
        items = new ArrayList<T>(rows * cols);
    }

    public T get(int row, int col) {
        return items.get(row * cols + col);
    }

    public void set(int row, int col, T value) {
        items.set(rows * cols + col, value);
    }

    @Override
    public String toString() {
        return "A grid with " + rows + " rows and " + cols + " columns";
    }
}

File: src/main/kotlin/_03_staying_immutable/generic/Grid.kt
package _03_staying_immutable.generic

class Grid<T>(val rows: Int, val cols: Int, initialValue: T) {

    private val items = MutableList<T>(size = rows * cols) {
        initialValue
    }

    fun get(row: Int, col: Int): T {
        return items[row * cols + col]
    }

    fun set(row: Int, col: Int, value: T) {
        items[rows * cols + col] = value
    }

    override fun toString(): String {
        return "A grid with $rows rows and $cols columns"
    }
}